File: Python/01_Strings_Arrays/ZeroMatrix.py
import matrix_utils

def clear_row(mat, N, M, row):
    for j in range(0, M):
        mat[row][j] = 0
    return


def clear_col(mat, N, M, col):
    for i in range(0, N):
        mat[i][col] = 0;
    return


def zero_matrix(mat, N, M):
    row_count = set()
    col_count = set()

    for i in range(0, N):
        for j in range(0, M):
            if mat[i][j] == 0:
                row_count.add(i)
                col_count.add(j)
    for row in row_count:
        clear_row(mat, N, M, row)
    for col in col_count:
        clear_col(mat, N, M, col)
    return


def zero_matrix_2(mat, N, M):
    clear_row_0 = False
    clear_col_0 = False

    for i in range(0, N):
        if mat[i][0] == 0:
            clear_col_0 = True
            break

    for j in range(0, M):
        if mat[0][j] == 0:
            clear_row_0 = True
            break

    for i in range(0, N):
        for j in range(0, M):
            if mat[i][j] == 0:
                mat[0][j] = 0
                mat[i][0] = 0

    for i in range(1, N):
        if mat[i][0] == 0:
            clear_row(mat, N, M, i)

    for j in range(1, M):
        if mat[0][j] == 0:
            clear_col(mat, N, M, j)

    if clear_row_0:
        clear_row(mat, N, M, 0)
        
    if clear_col_0:
        clear_col(mat, N, M, 0)
    return

File: C++/Practice_Pad/source/CH1/zero_matrix.cpp
#include "pch.h"
#include "zero_matrix.h"
#include "matrix_utils.h"

#include <cstdbool>

void clear_row(int ** mat, int N, int M, int row) {
    for (int j = 0; j < M; j++) {
        mat[row][j] = 0;
    }
}

void clear_col(int ** mat, int N, int M, int col) {
    for (int i = 0; i < N; i++) {
        mat[i][col] = 0;
    }
}
// O(n^2) time O(N+M) space
void zero_matrix(int** mat, int N, int M)
{
    bool *count_rows = new bool[N];
    bool *count_cols = new bool[M];
    for (int i = 0; i < N; i++) {
        count_rows[i] = false;
    }
    for (int j = 0; j < M; j++) {
        count_cols[j] = false;
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (mat[i][j] == 0) {
                count_rows[i] = true;
                count_cols[j] = true;
            }
        }
    }
    for (int i = 0; i < N; i++) {
        if (count_rows[i]) {
            clear_row(mat, N, M, i);
        }
    }
    for (int j = 0; j < M; j++) {
        if (count_cols[j]) {
            clear_col(mat, N, M, j);
        }
    }
}

// O(n^2) time O(1) space
void zero_matrix_2(int** mat, int N, int M) {
    bool clear_row_0 = false;
    bool clear_col_0 = false;

    for (int i = 0; i < N; i++) {
        if (mat[i][0] == 0) {
            clear_col_0 = true;
            break;
        }
    }
    for (int j = 0; j < M; j++) {
        if (mat[0][j] == 0) {
            clear_row_0 = true;
            break;
        }
    }
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (mat[i][j] == 0){
                mat[0][j] = 0;
                mat[i][0] = 0;
            }
        }
    }
    for (int i = 1; i < N; i++) {
        if (mat[i][0] == 0) {
            clear_row(mat, N, M, i);
        }
    }
    for (int j = 1; j < M; j++) {
        if (mat[0][j] == 0) {
            clear_col(mat, N, M, j);
        }
    }
    if (clear_col_0)
        clear_col(mat, N, M, 0);

    if (clear_row_0)
        clear_row(mat, N, M, 0);
}